tpl/images/images.go

```diff
@@ -117,3 +124,60 @@ func (ns *Namespace) Filter(args ...any) (images.ImageResource, error) {
 
 	return img.Filter(filtersv...)
 }
+
+type qrOptions struct {
+	Text      string // text to encode
+	Level     string // error correction level; one of low, medium, quartile, or high
+	TargetDir string // target directory relative to publishDir
+}
+
+var qrErrorCorrectionLevels = map[string]qr.Level{
+	"low":      qr.L,
+	"medium":   qr.M,
+	"quartile": qr.Q,
+	"high":     qr.H,
+}
+
+const qrDefaultErrorCorrectionLevel = "medium"
+
+// QR encodes text to a QR code using the given error correction level,
+// returning an image resource.
+func (ns *Namespace) QR(opts any) (images.ImageResource, error) {
+	qrOptions := qrOptions{
+		Level: qrDefaultErrorCorrectionLevel,
+	}
+
+	err := mapstructure.WeakDecode(opts, &qrOptions)
+	if err != nil {
+		return nil, err
+	}
+
+	if qrOptions.Text == "" {
+		return nil, errors.New("cannot encode an empty string")
+	}
+
+	level, ok := qrErrorCorrectionLevels[qrOptions.Level]
+	if !ok {
+		return nil, errors.New("error correction level must be one of low, medium, quartile, or high")
+	}
+
+	code, err := qr.Encode(qrOptions.Text, level)
+	if err != nil {
+		return nil, err
+	}
+
+	hash := hashing.XxHashFromStringHexEncoded(qrOptions.Text + qrOptions.Level)
+	targetPath := filepath.Join(qrOptions.TargetDir, fmt.Sprintf("qr_%s.png", hash))
```

Comment:
This all looks good now. Just one tiny detail: FromString expects Unix paths; I'm assuming that Windows paths will somehow work as well, but I suggest that you change the above to use `path.Join...`.

Reply:
Will do.

I'm not thrilled about what you currently need to do to scale these down to something other that [8 image pixels per QR code module](https://github.com/rsc/qr/blob/master/qr.go#L77). See last example in documentation. So I may add another option if I'm satisfied with the result. Great idea to pass a map instead of scalars... I get it. It makes additions almost trivial.

Reply:
Yea, going forward, I see 3 variants for template funcs args:

1. One single argument.
2. Multiple arguments where one of them is a obvious candidate for "piping", e.g. `{{ $resource | js.Build (dict "other" "option" }}`
3. Multiple arguments where there's no obvious pipe candidate: One options map.